### PR TITLE
Add repo superclass and take out methods in individual repos that are unneeded

### PR DESCRIPTION
**Pull request by avjohnston**
Added the repo superclass and then took out all the methods that we we reusing in our original repos, then fixed the tests for those files.

### CODE REVIEW

**Comment by brisag** [comment]
I like the inheriting! Thanks for clarifying modules vs inheritance for this one

File: lib/transaction_repository.rb [hunk 1]
```diff
@@ -47,15 +35,15 @@ def find_all_by_result(result)
...
+  def max_by_id
     @transactions.max_by do |transaction|
       transaction.id
     end.id
   end
 
   def create(attributes)
     @transactions.push(Transaction.new({
-                          id: (max_transaction_id + 1),
+                          id: (max_by_id + 1),
```
[the remark this answers is not in the document]
Comment (brisag) [reply lib/transaction_repository.rb:34]: small indents to make it consistent!

**Review by Diana20920 — APPROVED** [review]
Houd is just pointing out a couple extra blank lines, but otherwise this looks really good!! 🔥🔥

File: lib/invoice_item_repository.rb [hunk 2]
```diff
@@ -41,15 +29,15 @@ def find_all_by_invoice_id(invoice_id)
     end
   end
```
Comment (DougWelchons) [inline lib/invoice_item_repository.rb]: find_all_by_item_id(item_id) and find_all_by_invoice_id(invoice_id) show up in more then one class. so we should probably move them to the repository super class

File: lib/invoice_item_repository.rb [hunk 3]
```diff
@@ -41,15 +29,15 @@ def find_all_by_invoice_id(invoice_id)
...
 
-  def max_invoice_item_id
+  def max_by_id
     @invoice_items.max_by do |ii|
       ii.id
     end.id
   end
```
Comment (DougWelchons) [inline lib/invoice_item_repository.rb]: This has been moved to the super class. we should be able to remove this method from the class

File: lib/invoice_repository.rb [hunk 4]
```diff
@@ -47,15 +35,15 @@ def find_all_by_status(status)
     end
   end
 
-  def max_invoice_id
+  def max_by_id
     @invoices.max_by do |invoice|
       invoice.id
     end.id
   end
```
Comment (DougWelchons) [inline lib/invoice_repository.rb]: Same here. already in the super class.

File: lib/invoice_repository.rb [hunk 5]
```diff
@@ -47,15 +35,15 @@ def find_all_by_status(status)
     end
   end
```
Comment (DougWelchons) [inline lib/invoice_repository.rb:31]: I don't know why it wont let me comment on lines 26-30 but the find_all_by_merchant_id(merchant_id) method is duplicated, and can be moved to the super class.

File: lib/item_repository.rb [hunk 6]
```diff
@@ -59,15 +47,15 @@ def find_all_by_merchant_id(merchant_id)
     end
```
Comment (DougWelchons) [inline lib/item_repository.rb]: Same here, wont let me comment, but line 44-48 find_all_by_merchant_id(merchant_id), is doubled up.

File: lib/item_repository.rb [hunk 7]
```diff
@@ -59,15 +47,15 @@ def find_all_by_merchant_id(merchant_id)
...
   end
 
-  def max_item_id
+  def max_by_id
     @items.max_by do |item|
       item.id
     end.id
   end
```
Comment (DougWelchons) [inline lib/item_repository.rb]: can be removed

File: lib/merchant_repository.rb [hunk 8]
```diff
@@ -16,19 +18,19 @@ def create_repository(file_path)
...
+  #
+  # def all
+  #   @merchants
+  # end
+  #
+  # def find_by_id(id)
+  #   @merchants.find do |merchant|
+  #     merchant.id.to_i == id
+  #   end
+  # end
```
Comment (DougWelchons) [inline lib/merchant_repository.rb]: we should remove these commented out lines.

File: lib/merchant_repository.rb [hunk 9]
```diff
@@ -42,15 +44,15 @@ def find_all_by_name(name)
     end
   end
 
-  def max_merchant_id
+  def max_by_id
     @merchants.max_by do |merchant|
       merchant.id
     end.id
   end
```
Comment (DougWelchons) [inline lib/merchant_repository.rb]: ditto.

File: lib/merchant_repository.rb [hunk 10]
```diff
@@ -63,7 +65,7 @@ def update(id, attribute)
     find_by_id(id).updated_at = Time.now.round
   end
 
-  def delete(id)
-    @merchants.delete(find_by_id(id))
-  end
+  # def delete(id)
+  #   @merchants.delete(find_by_id(id))
+  # end
```
Comment (DougWelchons) [inline lib/merchant_repository.rb]: same here we should re move the commented out lines

File: lib/transaction_repository.rb [hunk 11]
```diff
@@ -15,20 +17,6 @@ def create_repository(file_path)
...
-
-  def find_by_id(id)
-    @transactions.find do |transaction|
-      transaction.id.to_i == id
-    end
-  end
-
   def find_all_by_invoice_id(id)
     @transactions.find_all do |transaction|
       transaction.invoice_id == id
```
Comment (DougWelchons) [inline lib/transaction_repository.rb]: find_all_by_invoice_id(id) is doubled up

File: lib/transaction_repository.rb [hunk 12]
```diff
@@ -47,15 +35,15 @@ def find_all_by_result(result)
     end
   end
 
-  def max_transaction_id
+  def max_by_id
     @transactions.max_by do |transaction|
       transaction.id
     end.id
   end
```
Comment (DougWelchons) [inline lib/transaction_repository.rb]: should be removed.

**Review by DougWelchons — COMMENTED** [review]
I should have been paying better attention while we are making these changes, we should have been removing the max_by_id methods instead of modifying it.
also we have a handful of other methods that are doubled up in our repository classes that should be moved to our super class.
lastly, we forgot to remove the commented out lines in the merchant repository class.

**Comment by brisag** [comment]
Thanks for moving all the find_all to the super class

File: lib/repository.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def find_by_id(id)
+    @repo.find do |instance|
+      instance.id.to_i == id
+    end
+  end
+
+  def max_by_id
+    @repo.max_by do |instance|
```
[the remark this answers is not in the document]
Comment (brisag) [reply lib/repository.rb:17]: I dont think we need to worry about it

**Comment by DougWelchons** [comment]
Looks good 👍

**Pull request merged**